cluster/src/main/java/io/scalecube/cluster/gossip/GossipProtocolImpl.java

```diff
@@ -179,31 +181,59 @@ private void doSpreadGossip() {
   // ================================================
 
   private String createAndPutGossip(Message message) {
-    long period = this.currentPeriod;
-    Gossip gossip = new Gossip(generateGossipId(), message);
-    GossipState gossipState = new GossipState(gossip, period);
+    final long period = this.currentPeriod;
+    final Gossip gossip = createGossip(message);
+    final GossipState gossipState = new GossipState(gossip, period);
+
     gossips.put(gossip.gossipId(), gossipState);
+    ensureSequence(localMember.id()).add(gossip.sequenceId());
+
     return gossip.gossipId();
   }
 
   private void onGossipReq(Message message) {
-    long period = this.currentPeriod;
-    GossipRequest gossipRequest = message.data();
+    final long period = this.currentPeriod;
+    final GossipRequest gossipRequest = message.data();
     for (Gossip gossip : gossipRequest.gossips()) {
-      GossipState gossipState = gossips.get(gossip.gossipId());
-      if (gossipState == null) { // new gossip
-        gossipState = new GossipState(gossip, period);
-        gossips.put(gossip.gossipId(), gossipState);
-        sink.next(gossip.message());
+      final SequenceIdCollector sequenceIdCollector = ensureSequence(gossip.gossiperId());
+      final boolean notExistBefore = sequenceIdCollector.add(gossip.sequenceId());
+
+      healthCheck(gossip.gossiperId(), gossip.message().sender(), sequenceIdCollector);
+
+      if (notExistBefore) {
+        GossipState gossipState = gossips.get(gossip.gossipId());
+        if (gossipState == null) { // new gossip
+          gossipState = new GossipState(gossip, period);
+          gossips.put(gossip.gossipId(), gossipState);
+          sink.next(gossip.message());
+        }
+        gossipState.addToInfected(gossipRequest.from());
       }
-      gossipState.addToInfected(gossipRequest.from());
+    }
+  }
+
+  private void healthCheck(
```

Comment:
This function is laying on the _hot-path_ of the algorithm. Incoming gossip traffic is being handled on it. IMO it's better to move this function to some more quite place .. e.g.  make it a part of   `doSpreadGossip()`.